### ISSUE TITLE
Alpine + Docker: Segfault with Input CPU and Output Fluentd

### ISSUE THREAD

**Issue by mirosval** [fluent/fluent-bit#223]
Using the default alpine-based fluentd image and fluent-bit image from the [here](https://github.com/fluent/fluent-bit-docker-image) and the following configuration, I get a segfault in fluent bit:

```zsh
Creating network "fluentbit_default" with the default driver
Creating fluentbit_fluentd_1
Creating fluentbit_fluentbit_1
Attaching to fluentbit_fluentd_1, fluentbit_fluentbit_1
fluentd_1    | 2017-04-03 15:52:00 +0000 [info]: reading config file path="/fluentd/etc/fluent.conf"
fluentbit_1  | Fluent-Bit v0.11.0
fluentd_1    | 2017-04-03 15:52:00 +0000 [info]: starting fluentd-0.12.32
fluentbit_1  | [2017/04/03 15:52:01] [ info] [engine] started
fluentd_1    | 2017-04-03 15:52:01 +0000 [info]: gem 'fluent-plugin-elasticsearch' version '1.9.3'
fluentd_1    | 2017-04-03 15:52:01 +0000 [info]: gem 'fluent-plugin-record-reformer' version '0.9.0'
fluentd_1    | 2017-04-03 15:52:01 +0000 [info]: gem 'fluentd' version '0.12.32'
fluentd_1    | 2017-04-03 15:52:01 +0000 [info]: adding match pattern="**" type="stdout"
fluentd_1    | 2017-04-03 15:52:01 +0000 [info]: adding source type="forward"
fluentd_1    | 2017-04-03 15:52:01 +0000 [info]: using configuration file: <ROOT>
fluentd_1    |   <source>
fluentd_1    |     @type forward
fluentd_1    |     bind 0.0.0.0
fluentd_1    |     port 24224
fluentd_1    |   </source>
fluentd_1    |   <match **>
fluentd_1    |     @type stdout
fluentd_1    |   </match>
fluentd_1    | </ROOT>
fluentd_1    | 2017-04-03 15:52:01 +0000 [info]: listening fluent socket on 0.0.0.0:24224
fluentbit_fluentbit_1 exited with code 139
```

fluent-bit.conf
```conf
[SERVICE]
    Flush     1
    Daemon    Off
    Log_Level debug

[INPUT]
    Name        cpu
    Tag         cpu_usage

[OUTPUT]
    Name        forward
    Match       *
    Host        fluentd
    Port        24224
```

fluent.conf
```conf
<source>
  @type forward
  bind 0.0.0.0
  port 24224
</source>

<match **>
  @type stdout
</match>
```

docker-compose.yml
```yaml
version: '3'
services:
    fluentbit:
        build: fluentbit
        links:
            - fluentd
        volumes:
            - "./fluentbit/fluent-bit.conf:/fluent-bit/etc/fluent-bit.conf"
    fluentd:
        build: fluentd
        expose:
            - "24224"
        ports:
            - "24224:24224"
        volumes:
            - "./fluentd/fluent.conf:/fluentd/etc/fluent.conf"
```

It does not happen when the output in fluentbit is set to stdout, so I suspect it is the fluentbit-fluentd connection causing this. Further I have observed corruption of the messages in some cases, which leads me to believe that the problem is caused by reading beyond the end of an array. 

My hypothesis: this could be caused by how floats are serialized, because we observe floats with 1-4 decimal places so it could be that they are counted as constant width when in fact they are variable width.

Please let me know if you need further info.

**Comment by edsiper** [comment]
troubleshooting.

**Comment by edsiper** [comment]
I was able to reproduce the issue with Docker Compose (Alpine Linux Docker images): 

- I cannot reproduce from my host OS

Now trying:

- Newer Alpine Linux Image (3.5)
- Enabling Valgrind to debug the problem

**Comment by edsiper** [comment]
It looks like an issue with Musl. The big problem is that Valgrind is not working properly:

```
/ # valgrind fluent-bit/bin/fluent-bit -i cpu -o stdout -m '*' -o stdout -m '*'
==8== Memcheck, a memory error detector
==8== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==8== Using Valgrind-3.12.0 and LibVEX; rerun with -h for copyright info
==8== Command: fluent-bit/bin/fluent-bit -i cpu -o stdout -m * -o stdout -m *
==8== 
==8== Invalid free() / delete / delete[] / realloc()
==8==    at 0x4C939EA: free (vg_replace_malloc.c:530)
==8==    by 0x405792D: ??? (in /lib/ld-musl-x86_64.so.1)
==8==  Address 0x4e9b180 is in a rw- mapped file /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so segment
==8== 
Fluent-Bit v0.11.0
Copyright (C) Treasure Data

==8== Thread 2:
==8== Syscall param epoll_pwait(sigmask) points to unaddressable byte(s)
==8==    at 0x4020D09: epoll_pwait (in /lib/ld-musl-x86_64.so.1)
==8==    by 0x2D4FAD: mk_event_wait (mk_event.c:154)
==8==    by 0x12FFB3: log_worker_collector (flb_log.c:122)
==8==    by 0x13ED0E: step_callback (flb_worker.c:50)
==8==    by 0x40519D8: ??? (in /lib/ld-musl-x86_64.so.1)
==8==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==8== 
[2017/04/03 17:56:37] [ info] [engine] started
==8== Thread 1:
==8== Syscall param epoll_pwait(sigmask) points to unaddressable byte(s)
==8==    at 0x4020D09: epoll_pwait (in /lib/ld-musl-x86_64.so.1)
==8==    by 0x2D4FAD: mk_event_wait (mk_event.c:154)
==8==    by 0x13A5F5: flb_engine_start (flb_engine.c:448)
==8==    by 0x12FC5F: main (fluent-bit.c:675)
==8==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==8== 
==8== Invalid read of size 1
==8==    at 0x2D6615: msgpack_object_bin_print (objectc.c:123)
==8==    by 0x2D6615: msgpack_object_print (objectc.c:210)
==8==    by 0x2D691D: msgpack_object_print (objectc.c:251)
==8==    by 0x2D69C5: msgpack_object_print (objectc.c:234)
==8==    by 0x170338: cb_stdout_flush (stdout.c:54)
==8==    by 0x13AE8E: output_pre_cb_flush (flb_output.h:304)
==8==    by 0x1ED386: co_init (amd64.c:117)
==8==  Address 0x4ea9cf9 is 25 bytes inside a block of size 1,308 alloc'd
==8==    at 0x4C929C5: malloc (vg_replace_malloc.c:299)
==8==    by 0x131982: flb_malloc (flb_mem.h:57)
==8==    by 0x133ACD: flb_input_flush (flb_input.c:825)
==8==    by 0x13BB1B: flb_engine_dispatch (flb_engine_dispatch.c:168)
==8==    by 0x139299: flb_engine_flush (flb_engine.c:82)
==8==    by 0x13A689: flb_engine_handle_event (flb_engine.c:256)
==8==    by 0x13A689: flb_engine_start (flb_engine.c:451)
==8==    by 0x12FC5F: main (fluent-bit.c:675)
==8== 
```

Since this is not the first time I see a problem with Musl I've decided to get rid of it, because:

- Fluent Bit ideally needs to be run with latest version of Jemalloc, Jemalloc + Musl are not friends, any app crashes.
- Cannot debug properly 

I will try to switch to debian gcr-0.1 image (lightweight debian from google) or another similar one that don't use Musl.

I will update the ticket once the changes are made.

**Comment by edsiper** [comment]
OK, so I just pushed a new image based on Debian image maintained by Google:

- https://github.com/fluent/fluent-bit-docker-image/blob/master/0.11/Dockerfile
- https://github.com/fluent/fluent-bit-docker-image

I did the test with docker-compose again and it worked properly, I was not longer able to reproduce the issue.

The new image will be available shortly on Docker Hub (Tag 0.11):

https://hub.docker.com/r/fluent/fluent-bit/tags/

Please let me know how it goes...

**Comment by edsiper** [comment]
I am closing this issue for now, if you face any anomaly feel free to re-open it or comment here.
